src/java.base/share/native/libjli/java.c

```diff
@@ -620,12 +614,32 @@ JavaMain(void* _args)
      * The main method is invoked here so that extraneous java stacks are not in
      * the application stack trace.
      */
-    if (!invokeStaticMainWithArgs(env, mainClass, mainArgs) &&
-        !invokeInstanceMainWithArgs(env, mainClass, mainArgs) &&
-        !invokeStaticMainWithoutArgs(env, mainClass) &&
-        !invokeInstanceMainWithoutArgs(env, mainClass)) {
-        ret = 1;
-        LEAVE();
+
+
+    helperClass = GetLauncherHelperClass(env);
+    isStaticMainField = (*env)->GetStaticFieldID(env, helperClass, "isStaticMain", "Z");
+    CHECK_EXCEPTION_NULL_LEAVE(isStaticMainField);
+    isStaticMain = (*env)->GetStaticBooleanField(env, helperClass, isStaticMainField);
+
+    noArgMainField = (*env)->GetStaticFieldID(env, helperClass, "noArgMain", "Z");
+    CHECK_EXCEPTION_NULL_LEAVE(noArgMainField);
+    noArgMain = (*env)->GetStaticBooleanField(env, helperClass, noArgMainField);
+
+    if (isStaticMain) {
+        if (noArgMain) {
+            ret = invokeStaticMainWithoutArgs(env, mainClass);
+        } else {
+            ret = invokeStaticMainWithArgs(env, mainClass, mainArgs);
+        }
+    } else {
+       if (noArgMain) {
+            ret = invokeInstanceMainWithoutArgs(env, mainClass);
+       } else {
+            ret = invokeInstanceMainWithArgs(env, mainClass, mainArgs);
+       }
```

Comment:
**Nit:** Wrong indentation[^1]:
```suggestion
        if (noArgMain) {
            ret = invokeInstanceMainWithoutArgs(env, mainClass);
        } else {
            ret = invokeInstanceMainWithArgs(env, mainClass, mainArgs);
        }
```

[^1]: (this is one of the reasons I hate when spaces are used to indent things)